src/main/java/com/microsoftopentechnologies/windowsazurestorage/service/DownloadService.java

```diff
@@ -32,49 +31,72 @@
 
 import java.io.IOException;
 import java.io.OutputStream;
-import java.net.URISyntaxException;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 
 public abstract class DownloadService extends StoragePluginService<DownloadServiceData> {
-
     protected static final String DOWNLOAD = "Download";
     protected static final String DOWNLOAD_FAILED = "DownloadFailed";
+    private static final int DOWNLOAD_THREAD_COUNT = 16;
+
+    private BlockingDeque<Object> downloadItemDeque = new LinkedBlockingDeque<>();
+    private int filesNeedDownload = 0;
+    private AtomicBoolean isScanFinished = new AtomicBoolean(false);
+    private AtomicInteger filesDownloaded = new AtomicInteger(0);
+    private Thread[] downloadThreads = new Thread[DOWNLOAD_THREAD_COUNT];
 
     public DownloadService(DownloadServiceData data) {
         super(data);
     }
 
-    protected int downloadBlobs(Iterable<ListBlobItem> blobItems)
-            throws URISyntaxException, StorageException, WAStorageException {
-        final DownloadServiceData serviceData = getServiceData();
-        int filesDownloaded = 0;
-        for (final ListBlobItem blobItem : blobItems) {
-            // If the item is a blob, not a virtual directory
-            if (blobItem instanceof CloudBlob) {
-                // Download the item and save it to a file with the same
-                final CloudBlob blob = (CloudBlob) blobItem;
-
-                // Check whether we should download it.
-                if (shouldDownload(
-                        serviceData.getIncludeFilesPattern(),
-                        serviceData.getExcludeFilesPattern(),
-                        blob.getName(),
-                        true)) {
-                    downloadBlob(blob);
-                    filesDownloaded++;
+    class DownloadThread implements Runnable {
+        @Override
+        public void run() {
+            while (!isScanFinished.get() || filesDownloaded.intValue() != filesNeedDownload) {
```

Comment:
Will the loop exit when something wrong which might cause `filesDownloaded.intValue() != filesNeedDownload` to be true forever?

Reply:
Delete this logic in new version.